invenio_rdm_records/links/models.py

```diff
@@ -0,0 +1,229 @@
+    )
+    """Creation timestamp."""
+
+    expires_at = db.Column(db.DateTime, nullable=True)
+    """Expiration date."""
+
+    revoked_at = db.Column(db.DateTime, nullable=True, index=True)
+    """Revocation timestamp."""
+
+    title = db.Column(db.String, nullable=False, default="")
+    """Title of link."""
+
+    description = db.Column(db.Text, nullable=False, default="")
+    """Description of link."""
+
+    permission_level = db.Column(db.String, nullable=False, default="")
+    """Permission level of the link."""
+
+    def get_absolute_url(self, endpoint):
+        """Get absolute for secret link (using https scheme).
+
+        Note: The token is passed to ``url_for`` as a keyword argument.
+        If the token has any data specified in its ``data`` dictionary,
+        they key-value pairs are also passed as keyword arguments.
+        """
+        copy = deepcopy(self.extra_data) or {}
+        if "recid" in copy:
+            copy["pid_value"] = copy.pop("recid")
+
+        return url_for(endpoint, token=self.token, _external=True, **copy)
+
+    def covers(self, permission):
+        """Check if the secret link covers the specified permission."""
+        # TODO permission hierarchy
+        return self.permission_level == permission
+
+    def revoke(self):
+        """Revoken a secret link."""
+        if self.revoked_at is None:
+            with db.session.begin_nested():
+                self.revoked_at = datetime.utcnow()
+
+            link_revoked.send(self)
+            return True
+
+        return False
+
+    @property
+    def need(self):
+        """Get the needs provided by the link."""
+        return LinkNeed(str(self.id))
+
+    @property
+    def extra_data(self):
+        """Load token data stored in token (ignores expiry date of tokens)."""
+        if self.token:
+            return self._load_token(self.token, force=True)["data"]
+
+        return None
+
+    @property
+    def is_expired(self):
+        """Determine if link is expired."""
+        if self.expires_at:
+            return datetime.utcnow() > self.expires_at
+
+        return False
+
+    @property
+    def is_revoked(self):
+        """Determine if link is revoked."""
+        return self.revoked_at is not None
+
+    @property
+    def is_valid(self):
+        """Determine if link is still valid."""
+        return not (self.is_expired or self.is_revoked)
+
+    @classmethod
+    def query_by_owner(cls, user):
+        """Get secret links by user."""
+        return cls.query.filter_by(owner_user_id=user.id)
+
+    @classmethod
+    def get_from_token(cls, token):
+        """Get the SecretLink referenced by the specified token."""
+        data = cls.load_token(token)
+        link = cls.query.get(data["id"])
+        return link
+
+    @classmethod
+    def create(
+        cls,
+        title,
+        owner,
+        permission_level,
+        extra_data,
+        description="",
+        expires_at=None,
+    ):
```

Comment:
that's gonna be the primary way of creating secret links, i think... along with `record.access.links.create(...)`